### PR TITLE
Add fused top-K softmax kernel for MoE

### PR DESCRIPTION
**Pull request by WoosukKwon**
This PR ports a fused topk-softmax kernel from TensorRT-LLM v0.7.1.

TODO:
1. Port more MoE-related kernels
2. Use CUTLASS-based grouped GEMM kernels with appropriate tuning (if they perform better than the current Triton kernel).

### CODE REVIEW

**Comment by WoosukKwon** [comment]
@Yard1 @cadedaniel @pcmoritz Can any one of you review the PR?

File: vllm/model_executor/layers/fused_moe.py [hunk 1]
```diff
@@ -210,12 +210,15 @@ def invoke_fused_moe_kernel(A: torch.Tensor, B: torch.Tensor, C: torch.Tensor,
...
-              inplace=False):
+def fused_moe(
+    hidden_states: torch.Tensor,
+    w1: torch.Tensor,
+    w2: torch.Tensor,
+    gating_output: torch.Tensor,
+    topk: int,
+    renormalize: bool,
+    inplace: bool = False,
+) -> torch.Tensor:
```
Comment (pcmoritz) [inline vllm/model_executor/layers/fused_moe.py:221]: The docstring for this function will need to be updated with the new parameters

File: vllm/model_executor/layers/fused_moe.py [hunk 2]
```diff
@@ -241,6 +244,38 @@ def fused_moe(hidden_states: torch.Tensor,
...
+    if is_hip():
+        # The MoE kernels are not yet supported on ROCm.
+        routing_weights = torch.softmax(gating_output,
+                                        dim=-1,
+                                        dtype=torch.float32)
+        topk_weights, topk_ids = torch.topk(routing_weights, topk, dim=-1)
+    else:
+        import vllm._moe_C as moe_kernels
+
+        num_tokens = hidden_states.numel() // hidden_states.shape[-1]
```
Comment (pcmoritz) [inline vllm/model_executor/layers/fused_moe.py:256]: It might also be slightly cleaner / easier to test if you refactor the code a little bit into something like this:
```python
if is_hip():
    topk_weights, topk_ids = fused_topk_torch(gating_output, topk)
else:
    topk_weights, topk_ids = fused_topk_cuda(gating_output, topk)
```

That would also address the problem that at the moment `M = num_tokens` in this function -- it is slightly strange to have two variables for the same thing :)

While at it, I would also add `assert gating_output.shape[0] == hidden_states.shape[0]` (for the number of tokens) above as well as `assert w1.shape[0] == gating_output.shape[1]` (for the number of expert) :)

File: csrc/moe/topk_softmax_kernels.cu [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+                softmax_workspace, nullptr, topk_weights, topk_indicies, token_expert_indices,
+                num_experts, topk, 0, num_experts);
+        }
+    }
+}
+
+} // namespace moe
+} // namespace vllm
+
+void topk_softmax(
```
Comment (pcmoritz) [inline csrc/moe/topk_softmax_kernels.cu:472]: The most natural argument ordering for me would be to put the `gating_output` parameter first since it is an IN parameter, and put all the other parameters after that since they are OUT parameters (also aligns better with the signature of `topkGatingSoftmaxKernelLauncher`).

Comment (WoosukKwon) [reply csrc/moe/topk_softmax_kernels.cu:472]: I think this is a matter of personal preference. For most custom ops in vLLM (which I wrote), I've listed the output parameters first and then listed the input parameters.

**Comment by pcmoritz** [comment]
Btw, I did a little bit of benchmarking on this PR and without touching any of the system parameters in the PR I'm already seeing a 1.5% - 3.5% latency improvement. It is higher in the low latency regime. Concretely I tested on TP2 on H100 with 1000 input and 50 output tokens on Mixtral. So it seems worth merging this even though the low-level kernel code is not easy to follow -- most people can probably just treat it as a black box so it shouldn't have a big impact on maintainability.

File: vllm/model_executor/layers/fused_moe.py (hunk 2 above)
Comment (pcmoritz) [reply vllm/model_executor/layers/fused_moe.py:256]: It would also be worth adding a small unit test of `topk_softmax_torch` vs. `fused_topk_softmax_cuda` in `test_moe.py` -- that will increase the maintainability of the code and make it easier to port to AMD going forward if that is deemed worthwhile.

File: csrc/moe/topk_softmax_kernels.cu (hunk 3 above)
Comment (pcmoritz) [reply csrc/moe/topk_softmax_kernels.cu:472]: Got it, makes sense!

**Review by pcmoritz — APPROVED** [review]
I will spend some more time trying to understand the implementation in `topk_softmax_kernels.cu` but no need to block on that since that's mostly the upstream code from `https://github.com/NVIDIA/TensorRT-LLM/blob/v0.7.1/cpp/tensorrt_llm/kernels/mixtureOfExperts/moe_kernels.cu` and in any case we should probably keep it close to that and not change it  :)

File: tests/kernels/test_moe.py [hunk 4]
```diff
@@ -51,10 +49,10 @@ def test_fused_moe(
...
 
     score = torch.randn((m, e), device='cuda', dtype=dtype)
-    score = torch.softmax(score, dim=-1)
-    topk_weight, topk_ids = torch.topk(score, topk)
+    triton_output = fused_moe(a, w1, w2, score, topk, renormalize=False)
 
-    triton_output = fused_moe(a, w1, w2, topk_weight, topk_ids, False)
+    score = torch.softmax(score, dim=-1, dtype=torch.float32)
+    topk_weight, topk_ids = torch.topk(score, topk)
     torch_output = torch_moe(a, w1, w2, topk_weight, topk_ids)
```
Comment (pcmoritz) [inline tests/kernels/test_moe.py:56]: One more thing I noticed: It might be worth pushing the
```
    score = torch.softmax(score, dim=-1, dtype=torch.float32)
    topk_weight, topk_ids = torch.topk(score, topk)
```
code into `torch_moe` so that `fused_moe` and `torch_moe` keep having the same signature and behavior :)

File: vllm/model_executor/layers/fused_moe.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/model_executor/layers/fused_moe.py:256]: Hi @pcmoritz, thanks for the review. The reason I didn't make a new function for topk_softmax is the chance that AMD and NVIDIA GPUs may have completely different implementations for MoE. Specifically, it is possible that AMD GPUs will use the current PyTorch + Triton while NVIDIA GPUs use the custom kernels ported from TRT-LLM. So, I was thinking that vLLM would eventually have 2 separate `fused_moe` implementations.

The reason is,
1. The custom kernels from TRT-LLM may not be portable to AMD GPUs. For example, the topk softmax kernel in this PR uses CUB, which to my knowledge cannot be simply translated via hipification. Other TRT-LLM kernels that will be added in the future PR are also not easy to port, since they use CUB or CUTLASS.
2. The two implementations are different in how it permutes/pads the input activations. For example, the TRT-LLM's topk softmax kernel returns `token_expert_indices` which will be used in subsequent permutation ops while the PyTorch-based implementation does not use it. Furthermore, TRT-LLM's grouped GEMM does not require padding between the tokens routed to different experts, while the current Triton kernel does.

Comment (pcmoritz) [reply vllm/model_executor/layers/fused_moe.py:256]: That makes sense to me! Then maybe just unify the `M` and `num_tokens` variables and add the asserts about `gating_output` to make the function a little more readable?

File: tests/kernels/test_moe.py (hunk 4 above)
Comment (WoosukKwon) [reply tests/kernels/test_moe.py:56]: Good point! Fixed.

File: vllm/model_executor/layers/fused_moe.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/model_executor/layers/fused_moe.py:221]: Good catch! Fixed.

File: vllm/model_executor/layers/fused_moe.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/model_executor/layers/fused_moe.py]: Sounds good! Added.

**Comment by WoosukKwon** [comment]
@pcmoritz Thanks again for you review! Yes, I think we don't have to worry too much about the implementation details, at least at the moment, as I only made a minor change to the kernel.

**Comment by pcmoritz** [comment]
Sounds good, the PR looks great :)
